openhtf/plugs/usb/__init__.py

```diff
@@ -36,56 +38,153 @@ def MyPhase(test, adb):
+usb_hub: local
+
+---- Config file example for android/fastboot usb at local
+device_type: general
+usb_hub: 
+   local:
+      class: xxxxx
+      subclass: xxxxx
+      protocol: xxxxx
+      vendor_id:  xxxx (optional)
+      product_id: xxxxx(optional)
+      serial: (device serial number, optional)
+"""
 
 
-def _open_usb_handle(**kwargs):
-  """Open a UsbHandle subclass, based on configuration.
-
-  If configuration 'usb_server' is set, use it to connect to remote usb,
-  otherwise attempt to connect locally.
-
-  Args:
-    **kwargs: Arguments to pass to respective handle's Open() method.
-
-  Returns:
-    Instance of UsbHandle.
-  """
-  if conf.Config().usb_server:
-    return None
+def _get_usb_serial(port_addr):
+  """Get a usb serial based on the Cambrionix unit mac address in configuration."""
+  cmd = '/usr/local/google/home/amyxchen/esuit64 -t "DEVICE INFO, %s"' % port_addr
+  info = commands.getstatusoutput(cmd)[1]
+  serial_info = info.split('SERIAL:')[1]
+  serial = serial_info.split('\n')[0].strip()
+
+  _LOG.info('get serial:%s on port:%s' % (serial, port_addr))
+
+  return serial
+
+def _open_usb_handle():
+  dev_type = conf.Config().device_type
+  usb_hub = conf.Config().usb_hub
+  serial = None
+  vendor_id = None
+  product_id = None
+  device = None
+
+  if isinstance(usb_hub, dict):
+    if usb_hub.has_key('Ethersync'):
+      device = usb_hub['Ethersync']
+      if isinstance(device, dict) and device.has_key('port_addr'):
+        serial = _get_usb_serial(device['port_addr'])
+      else:
+        raise GeneralUsbAttributeError('Ethersync needs port_addr address and and/or mac_addr')
+    else:
+      device = usb_hub['local']
   else:
-    return local_usb.LibUsbHandle.Open(**kwargs)
-
-
+    device = usb_hub
+
+  if dev_type == 'fastboot':
+    device = fastboot_device
+  elif dev_type == 'android':
+    device = adb_device
+  elif dev_type == 'general':
+    if isinstance(device, dict):
+      try:
+        vendor_id = device['vendor_id']
+      except KeyError:
+        pass
+      try:
+        product_id = device['product_id']
+      except KeyError:
+        pass
+      try:
+        serial = device['serial']
+      except KeyError:
+        pass
+
+      try:
+        device = general_usb(device['class'], device['subclass'], device['protocol'])
+      except KeyError:
+        raise GeneralUsbAttributeError('must set class\subclass\protocol for \
+                                        general type device:%s', device)
+    else:
+      raise GeneralUsbAttributeError('must set attribute for general \
+                                      device:%s', device)
+  else:
+    raise GeneralUsbAttributeError('must set device type')
+
+  return local_usb.LibUsbHandle.Open(
+         interface_class=device.CLASS,
+         interface_subclass=device.SUBCLASS,
+         interface_protocol=device.PROTOCOL,
+         serial_number=serial,
+         vendor_id=vendor_id,
+         product_id=product_id)
+ 
```

Comment:
Per our in-person discussion, the individual plug code seems like the correct place to introduce constraints specific to that plug. Especially given that a single test may use multiple forms of USB connection before it's done.